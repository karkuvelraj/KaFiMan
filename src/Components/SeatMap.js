import React ,{useEffect, useState} from "react"
import {connect }from "react-redux"
import * as passengerActions from "../redux/actions/passengerActions"
import PropTypes from "prop-types"
import { toast } from 'react-toastify';
import Tooltip from '@material-ui/core/Tooltip';
import {bindActionCreators} from 'redux'

export function SeatMap(props){


    const seats=[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]
    let passengerMap={}
    if(props.passengers!=null)
    props.passengers.forEach((p)=>{
        passengerMap[p.seatNo]=p
    })
    const updateCheckin=(passenger)=>{
             if(passenger.isCheckedIn)
             toast(`${passenger.name} Marked as not Checked In`)
            else
            toast(`${passenger.name} Checked In`)
        props.updatePassenger({...passenger,isCheckedIn:!passenger.isCheckedIn})
    }
    const [opts,setOpts] =useState("checkedIn")
    
    useEffect(()=>{
        if(props.mealPrefer)
        setOpts("meals")
    },[])

    const radioHandle=({target})=>{
        setOpts(target.value)
     }
     const colorCheck=(passenger)=>{
        if(opts=="checkedIn")return passenger.isCheckedIn
        if(opts=="wheel")return passenger.isWheelChair
        if(opts=="infant")return passenger.isInfant
        if(opts=="meals")return passenger.mealPrefer
     }
    //  let label={"checkedIn":"Checked In","wheel": "Wheel Chair","infant":"Infant","meals":"Special Meals"}
     let selectSeat=(x)=>{
         if(props.admin)
         if(!passengerMap[x]){
             
        toast(`Seat Number ${x} selected.`)
         props.select(x)
         }else
         toast(`Seat Number ${x} Already Booked. Select Empty Seat`)
     }
    return (<>
    <div className="head">
        <span className="heading">Seat Map</span>
        <div style={{paddingLeft:'20px'}}>
        {props.mealPrefer?null:props.admin?<span id="maptype">Select Unbooked Seat </span>: <>
                 <label htmlFor="checkedInS">CheckedIn</label>
                 <input type="radio" name="checkedInS" id="type" value="checkedIn" checked={opts=="checkedIn"} onChange={radioHandle} />
                 <label htmlFor="wheelS">Wheel Chair</label>
                 <input type="radio" name="wheelS" id="type" value="wheel" checked={opts=="wheel"} onChange={radioHandle} />
                 <label htmlFor="infantS">Infant</label>
                 <input type="radio" name="infantS" id="type" value="infant"checked={opts=="infant"} onChange={radioHandle} />
                 </>}
                 {opts=='meals'?<span id="maptype">Passengers with Special meals preference</span>:null}
                 </div>
    </div>
    <div className="first" >
       {seats.map((x,i)=>{
                return !passengerMap[x]?<Tooltip key={i} title='Empty Seat'> 
                            <div className="seat empty"  onClick={()=>selectSeat(x)} >
                                 <p>Seat No: <span className="seatNo">{x}</span></p>
                                 </div>
                                 </Tooltip>
                                 :
                                <Tooltip key={i} title={
                                    <span className="tooltip">
                                    <p>Name : {passengerMap[x].name}</p>
                                    {passengerMap[x].isCheckedIn?<p>Checked In</p>:null}
                                    {passengerMap[x].isWheelChair?<p>Wheelchair Assistance</p>:null}
                                    {passengerMap[x].isInfant?<p>Infant</p>:null}
                                </span>
                                }>
                            <div className={colorCheck(passengerMap[x])?"checkedin seat":"seat"} 
                                onClick={()=>selectSeat(x)}
                            key={i} >
                              
                                <p>Seat No: <span className="seatNo">{x}</span></p>
                                <p>{passengerMap[x].name}</p>
                                {/* <p>{passengerMap[x].isCheckedIn?"Checked In":""}</p> */}
                               {props.mealPrefer||props.admin?null: <button className="s-btn" style={{height:'30px'}}
                                onClick={()=>updateCheckin(passengerMap[x])}>
                                    {passengerMap[x].isCheckedIn?"Undo":"CheckIn"}</button>}
                                </div>
                                </Tooltip>
                        })
                        }
                    </div>
    </>)
}
SeatMap.propTypes={
    passengers:PropTypes.array.isRequired,
    updatePassenger:PropTypes.func.isRequired,
    mealPrefer:PropTypes.any,
    admin:PropTypes.any,
    select:PropTypes.func
}
const mapToProps =({passengers})=>{
    return {passengers}
}
const mapActions=(dispatch)=>{
    return {...bindActionCreators(passengerActions,dispatch)}
}
export default connect(mapToProps,mapActions)(SeatMap)