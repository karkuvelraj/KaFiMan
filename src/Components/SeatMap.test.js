import React from "react"
import {shallow,mount} from "enzyme"
import {SeatMap} from "./SeatMap"
global.fetch = jest.fn(() =>
  Promise.resolve({json: () => Promise.resolve({ })}),
  Promise.reject({json: () => Promise.reject({ }),})
);   
let passengers=[
    {
        id:"ASD1000",
        name: "Vimal raja",
        dob: "1996-04-05",
        passportNo: "ASD1000",
        seatNo: "4",
        flightNo:"F10W10",
        address: "",
        isCheckedIn: true,
        isWheelChair: false,
        isInfant: false,
        services: [1,2,4],
        mealPrefer: "",
        shopItems: [2,1,4]
      },
      {
        id:"FDD1888",
        name: "Arun raja",
        dob: "",
        passportNo: "FDD1888",
        seatNo: "3",
        flightNo:"F10W10",
        address: "51,1st street,chennai",
        isCheckedIn: true,
        isWheelChair: false,
        isInfant: false,
        services: [1,2,4],
        mealPrefer: "BLML",
        shopItems: [2,1,4]
      }
]
it('rendering SearMap in InFlight',()=>{
    const wrapper=shallow(<SeatMap passengers={passengers} updatePassenger={()=>{}} mealPrefer/>)
    expect(wrapper).toMatchSnapshot()
})

it('rendering SearMap in checkIn',()=>{
    const wrapper=shallow(<SeatMap passengers={passengers} updatePassenger={()=>{}}/>)
    expect(wrapper).toMatchSnapshot()
})

it('rendering SearMap For Admin',()=>{
    const wrapper=shallow(<SeatMap passengers={passengers} admin updatePassenger={()=>{}}/>)
    expect(wrapper).toMatchSnapshot()
})

it('should distinguish checkedIn from UnChecked',()=>{
    
    const wrapper=shallow(<SeatMap passengers={passengers} updatePassenger={()=>{}} />)
    let checkedInpassengers=wrapper.find('.checkedin')
    expect(checkedInpassengers.length).toBe(2)

})

it('should display undo checkin button ',()=>{
    
    const wrapper=shallow(<SeatMap passengers={passengers} updatePassenger={()=>{}} />)
    let button=wrapper.find('.checkedin').first().find("button")
  
    expect(button.exists()).toEqual(true)
})

it('should not display checkin button Inflight',()=>{
    
    const wrapper=shallow(<SeatMap passengers={passengers} updatePassenger={()=>{}} mealPrefer/>)
    let button=wrapper.find('.checkedin').first().find("button")
  
    expect(button.exists()).toEqual(false)
})
it('click undo passenger check in',()=>{
    let updateAction=jest.fn()
    const wrapper=mount(<SeatMap passengers={passengers} updatePassenger={updateAction} />)
    wrapper.find('.checkedin').first().find("button").simulate('click')
    
    expect(updateAction).toBeCalled()
    expect(updateAction).toBeCalledWith({...passengers[1],isCheckedIn:!passengers[1].isCheckedIn})
})

it('Admin select first empty seat for new passengers',()=>{
    let select=jest.fn()
    let updateAction=jest.fn()
    const wrapper=mount(<SeatMap passengers={passengers} admin select={select} updatePassenger={updateAction} />)
    wrapper.find('.empty').first().simulate('click')
    
    expect(select).toBeCalled()
    expect(select).toBeCalledWith(1)
})

it('should  add heading in Inflight Seat Map',()=>{
    
    const wrapper=mount(<SeatMap passengers={passengers} updatePassenger={()=>{}} mealPrefer/>)
    
  
    expect(wrapper.find(".head").find('div').last()
        .find('#maptype').text()).toEqual('Passengers with Special meals preference')
})