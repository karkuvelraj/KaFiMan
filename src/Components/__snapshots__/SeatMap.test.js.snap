// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering SearMap For Admin 1`] = `ShallowWrapper {}`;

exports[`rendering SearMap in InFlight 1`] = `ShallowWrapper {}`;

exports[`rendering SearMap in checkIn 1`] = `ShallowWrapper {}`;
